packages/foundry/contracts/other/tether/TetherJoin.sol

```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity >=0.8.13;
+
+import "./IUSDT.sol";
+import "../../FlashJoin.sol";
+import "@yield-protocol/utils-v2/contracts/token/IERC20.sol";
+import "@yield-protocol/utils-v2/contracts/access/AccessControl.sol";
+import "@yield-protocol/utils-v2/contracts/token/TransferHelper.sol";
+
+contract TetherJoin is FlashJoin {
+    using TransferHelper for IERC20;
+
+    constructor(address asset_) FlashJoin(asset_) {}
+
+    /// @dev Calculate the amount of `asset` that needs to be sent to receive `amount` of USDT.
+    function _reverseFee(uint256 amount) internal view returns (uint256) {
+        return amount * 1000000 / (1000000 - (IUSDT(asset).basisPointsRate() * 100));
```

Comment:
I'm not sure of the math here.

The original code is:
`received = sent.sub(_value.mul(basisPointsRate).div(10000))`

So the reverse is:
```
received = sent - sent.mul(basisPointsRate).div(10000)
received = sent - sent *↓ (basisPointsRate /↓ 10000)
received = sent *↓ (1 - (basisPointsRate /↓ 10000))
received = sent *↓ (10000 - basisPointsRate)
sent = received /↑ (10000 - basisPointsRate)
```
Note that when we reverse a calculation, taking rounding into account, the regular *↓ becomes a /↑. The div(10000) is also a /↓ so I think it doesn't change the rounding.

I think that in your implementation the 1000000 and 100 factor changes are a bit confusing. To make implementation easiest I would import WDivUp from yield-utils and convert into a fixed point factor with 18 decimals.

`sent = received.wdivup(1e18 - basisPointsRate * 1e14)`

Reply:
Also, take into account that the fee is capped at `IUSDT(asset).maximumFee()`, so the above is:

`sent = min(received.wdivup(1e18 - basisPointsRate * 1e14), received + maximumFee)

Reply:
Will try to fix this.

Reply:
I implemented my own. ;)

Reply:
Actually `wdivup` will cause a rounding error and try to `transferFrom` one more wei than was approved. This is fixed with `wdiv`.

Reply:
In this example `2003807` is approved and `transferFrom`'d. `wdivup` will try to `transferFrom` `2003808`

![image](https://user-images.githubusercontent.com/24715302/215155983-9be49dee-84b4-4950-a86d-acbbee5cea36.png)